Importing libraries

In [31]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

importing dataset

In [6]:
df = pd.read_csv('SIH Bail Dec.  Dataset.csv')
df.shape

(500, 8)

In [7]:
df.head()

Type_of_Crime          Charges_Filed  Time_Served_Days  \
0       Violent                IPC 420              2549   
1   Non-violent                IPC 302              2188   
2       Violent  Economic Offences Act               111   
3         Cyber                IPC 420              1614   
4       Violent     Narcotic Drugs Act              2425   

   Prior_Criminal_History  Risk_of_Flight_Score Influence_on_Trial  \
0                       1              0.639427                NaN   
1                       0              0.025011                Low   
2                       1              0.275029               High   
3                       0              0.223211               High   
4                       1              0.736471               High   

  Bail_Decision Socio_Economic_Background  
0       Granted                    Middle  
1      Rejected                     Upper  
2       Granted                     Upper  
3       Granted                     Lower  
4      Rejected                     Lower

In [8]:
df.isna().sum()

Type_of_Crime                  0
Charges_Filed                  0
Time_Served_Days               0
Prior_Criminal_History         0
Risk_of_Flight_Score           0
Influence_on_Trial           142
Bail_Decision                  0
Socio_Economic_Background      0
dtype: int64

There are 142 rows with no entry in Influence _on _Trial column, so to handle them I will fill the empty values with 'nit' means No_Influence_on_Trial.

In [17]:
df['Influence_on_Trial'].fillna('nit', inplace =True)

In [10]:
df.isna().sum()

Type_of_Crime                0
Charges_Filed                0
Time_Served_Days             0
Prior_Criminal_History       0
Risk_of_Flight_Score         0
Influence_on_Trial           0
Bail_Decision                0
Socio_Economic_Background    0
dtype: int64

Now no column have missing values, so we are good to go.

In [11]:
df.duplicated().sum()

np.int64(0)

There are no duplicate values in the data set

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 8 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Type_of_Crime              500 non-null    object 
 1   Charges_Filed              500 non-null    object 
 2   Time_Served_Days           500 non-null    int64  
 3   Prior_Criminal_History     500 non-null    int64  
 4   Risk_of_Flight_Score       500 non-null    float64
 5   Influence_on_Trial         500 non-null    object 
 6   Bail_Decision              500 non-null    object 
 7   Socio_Economic_Background  500 non-null    object 
dtypes: float64(1), int64(2), object(5)
memory usage: 31.4+ KB


In [13]:

df.nunique()

Type_of_Crime                  6
Charges_Filed                  7
Time_Served_Days             478
Prior_Criminal_History         2
Risk_of_Flight_Score         500
Influence_on_Trial             4
Bail_Decision                  2
Socio_Economic_Background      3
dtype: int64

In [14]:
df.describe()

Time_Served_Days  Prior_Criminal_History  Risk_of_Flight_Score
count        500.000000              500.000000            500.000000
mean        1816.934000                0.540000              0.501442
std         1025.724174                0.498897              0.291295
min            5.000000                0.000000              0.000406
25%          984.500000                0.000000              0.247087
50%         1846.500000                1.000000              0.503814
75%         2676.000000                1.000000              0.752267
max         3644.000000                1.000000              0.999282

In [27]:
df.head()

Type_of_Crime          Charges_Filed  Time_Served_Days  \
0       Violent                IPC 420              2549   
1   Non-violent                IPC 302              2188   
2       Violent  Economic Offences Act               111   
3         Cyber                IPC 420              1614   
4       Violent     Narcotic Drugs Act              2425   

   Prior_Criminal_History  Risk_of_Flight_Score Influence_on_Trial  \
0                       1              0.639427                nit   
1                       0              0.025011                Low   
2                       1              0.275029               High   
3                       0              0.223211               High   
4                       1              0.736471               High   

  Bail_Decision Socio_Economic_Background  
0       Granted                    Middle  
1      Rejected                     Upper  
2       Granted                     Upper  
3       Granted                     Lower  
4      Rejected                     Lower

In [19]:
print("Categories in 'Type_of_Crime' variable:     ",end=" " )
print(df['Type_of_Crime'].unique())

print("Categories in 'Charges_Filed' variable:  ",end=" ")
print(df['Charges_Filed'].unique())

print("Categories in'parental level of education' Prior_Criminal_History:",end=" " )
print(df['Prior_Criminal_History'].unique())

print("Categories in 'Influence_on_Trial' variable:     ",end=" " )
print(df['Influence_on_Trial'].unique())

print("Categories in 'Bail_Decision' variable:     ",end=" " )
print(df['Bail_Decision'].unique())

print("Categories in 'Socio_Economic_Background' variable:     ",end=" " )
print(df['Socio_Economic_Background'].unique())

Categories in 'Type_of_Crime' variable:      ['Violent' 'Non-violent' 'Cyber' 'White-collar' 'Economic' 'Drug-related']
Categories in 'Charges_Filed' variable:   ['IPC 420' 'IPC 302' 'Economic Offences Act' 'Narcotic Drugs Act'
 'IPC 307' 'IPC 376' 'Cyber Crimes Act']
Categories in'parental level of education' Prior_Criminal_History: [1 0]
Categories in 'Influence_on_Trial' variable:      ['nit' 'Low' 'High' 'Moderate']
Categories in 'Bail_Decision' variable:      ['Granted' 'Rejected']
Categories in 'Socio_Economic_Background' variable:      ['Middle' 'Upper' 'Lower']


In [25]:
numeric_feature  = [feature for feature in df.columns if df[feature].dtype!= 'O']
categorical_feature = [feature for feature in df.columns if df[feature].dtype== 'O']
print('Numerical feature we have are: {}'.format( numeric_feature))
print('\nCategorical features we have are: {}'.format (categorical_feature))

Numerical feature we have are: ['Time_Served_Days', 'Prior_Criminal_History', 'Risk_of_Flight_Score']

Categorical features we have are: ['Type_of_Crime', 'Charges_Filed', 'Influence_on_Trial', 'Bail_Decision', 'Socio_Economic_Background']


In [36]:
df.to_csv('modified_bail_dataset.csv')

Feature Wise Visualization

Impact of Type_of_Crime on Bail_Decision

In [35]:
Type_of_Crime_group = df.groupby('Type_of_Crime')
Type_of_Crime_group

In [ ]:
plt.figure(figsize=(10,8))
X=['Bail_Decision']

